### PR TITLE
Adding requires_grad flag to Value class

### PR DESCRIPTION
**Pull request by rhalbersma**
I've added a `requires_grad` flag to the `Value` class that defaults to `False`, mimicking PyTorch's behavior. 

The topological sort of the computational graph can then stop expanding nodes that don't require gradient propagation. E.g. in expressions such as `x + 2` the conversion of the scalar `2` to `Value(2)` will no longer propagate the gradient for that node. 

I've also incorporated these changes into the neural network where all parameters receive `requires_grad=True`. All unit tests continue to pass and the demo notebook finds the exact same loss after 99 iterations.

Note that I also did some other small refactorings such as the separation between local gradient computation and propagation. This made it easier to selectively propagate the gradient based on the `requires_grad` flags. I've renamed the internal `_backward` attribute to a constructor argument `gradient_fn`.

All in all, the `engine.py` module was reduced by 15 lines, and weighs in at less than 80 lines now.

### CODE REVIEW

**Comment by rhalbersma** [comment]
Any update on this? Is this repo not taking PRs, or should I polish the PR a bit more?

**Comment by fcakyon** [comment]
@rhalbersma unfortunately this repo doesn't take PRs..
